compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -1017,22 +1007,39 @@ private STNode parseReturnTypeDescRhs(STNode metadata, STNode functionKeyword, S
         return funcDef;
     }
 
-    private STNode parseVarDeclWithFunctionType(STNode typeDesc, STNode qualifiers, STNode metadata,
-                                                boolean isObjectMember, boolean isObjectTypeDesc) {
+    private STNode parseVarDeclWithFunctionType(STNode functionKeyword, STNode funcSignature, STNode qualifiers,
+                                                STNode metadata, boolean isObjectMember, boolean isObjectTypeDesc) {
+        endContext(); // end the func-type
         STNodeList qualifierList = (STNodeList) qualifiers;
         STNode visibilityQualifier = STNodeFactory.createEmptyNode();
-        // qualifiers are not allowed except for object filed
+        List<STNode> validatedQualifierList = new ArrayList<>();
+        // qualifiers are only allowed in the following cases for func type desc.
+        // isolated qualifier allowed.
+        // public or private qualifier allowed in object field.
         for (int position = 0; position < qualifierList.size(); position++) {
             STNode qualifier = qualifierList.get(position);
-            if (isObjectMember && isVisibilityQualifier(qualifier)) {
+            if (qualifier.kind == SyntaxKind.ISOLATED_KEYWORD) {
+                validatedQualifierList.add(qualifier);
+            } else if (isObjectMember && isVisibilityQualifier(qualifier)) {
                 // public or private qualifier allowed in object field.
                 visibilityQualifier = qualifier;
             } else {
-                typeDesc = SyntaxErrors.cloneWithLeadingInvalidNodeMinutiae(typeDesc, qualifier,
+                functionKeyword = SyntaxErrors.cloneWithLeadingInvalidNodeMinutiae(functionKeyword, qualifier,
                         DiagnosticErrorCode.ERROR_QUALIFIER_NOT_ALLOWED, ((STToken) qualifier).text());
+
             }
         }
 
+        qualifiers = STNodeFactory.createNodeList(validatedQualifierList);
+        STNode typeDesc = STNodeFactory.createFunctionTypeDescriptorNode(qualifiers, functionKeyword,
+                funcSignature);
+
+        // Check if it is a complex type desc starting with function type
+        if (isValidTypeContinuationToken(peek())) {
```

Comment:
I think we don't need this check. `parseComplexTypeDescriptor` does the same check indirectly.

Reply:
I c thax ayye will change.